tests/unit/test_repository.py

```diff
@@ -0,0 +1,441 @@
+    path = "path 1"
+    branch = "branch 1"
+
+    returned_content = repository_client.get_file_content(path=path, branch=branch)
+
+    assert returned_content == content
+    mock_github_repository.get_contents.assert_called_once_with(path, branch)
+
+
+@pytest.mark.parametrize(
+    "remote_url",
+    [
+        pytest.param("https://gitlab.com/canonical/upload-charm-docs.git", id="non-github url"),
+        pytest.param("http://gitlab.com/canonical/upload-charm-docs.git", id="http url"),
+        pytest.param("git@github.com:yanksyoon/actionrefer.git", id="ssh url"),
+    ],
+)
+def test_get_repository_name_invalid(remote_url: str):
+    """
+    arrange: given a non-valid remote_url
+    act: when _get_repository_name is called
+    assert: InputError is raised.
+    """
+    with pytest.raises(InputError) as exc:
+        repository._get_repository_name_from_git_url(remote_url=remote_url)
+
+    assert_substrings_in_string(
+        ("invalid remote repository url",),
+        str(exc.value).lower(),
+    )
+
+
+@pytest.mark.parametrize(
+    "remote_url, expected_repository_name",
+    [
+        pytest.param(
+            "https://github.com/canonical/upload-charm-docs",
+            "canonical/upload-charm-docs",
+            id="valid url",
+        ),
+        pytest.param(
+            "https://github.com/canonical/upload-charm-docs.git",
+            "canonical/upload-charm-docs",
+            id="valid git url",
+        ),
+    ],
+)
+def test_get_repository_name(remote_url: str, expected_repository_name: str):
+    """
+    arrange: given a non-valid remote_url
+    act: when _get_repository_name is called
+    assert: GitError is raised.
+    """
+    assert (
+        repository._get_repository_name_from_git_url(remote_url=remote_url)
+        == expected_repository_name
+    )
+
+
+def test_create_repository_client_no_token(
+    repository_path: Path,
+):
+    """
+    arrange: given valid repository path and empty access_token
+    act: when create_repository_client is called
+    assert: InputError is raised.
+    """
+    # the following token is for testing purposes only.
+    test_token = ""  # nosec
+
+    with pytest.raises(InputError) as exc:
+        repository.create_repository_client(access_token=test_token, base_path=repository_path)
+
+    assert_substrings_in_string(
+        ("invalid", "access_token", "input", "it must be", "non-empty"),
+        str(exc.value).lower(),
+    )
+
+
+def test_create_repository_client(
+    monkeypatch: pytest.MonkeyPatch,
+    git_repo: Repo,
+    repository_path: Path,
+    mock_github_repo: Repository,
+):
+    """
+    arrange: given valid repository path and a valid access_token and a mocked github client
+    act: when create_repository_client is called
+    assert: RepositoryClient is returned.
+    """
+    origin = git_repo.remote("origin")
+    git_repo.delete_remote(origin)
+    git_repo.create_remote("origin", "https://github.com/test-user/test-repo.git")
```

Comment:
Why do you need git_repo and make changes here ?

Reply:
This was because the remote on the `git_repo` is local before the changes, which would throw an error when passed to the repository client. The line of code matching the repository is `repository.py:192` where origin url is matched against the regex pattern to get the branch name :)

Reply:
ok thanks, I'll add this as a comment

Reply:
Thank you!